pkg/oc/admin/migrate/scc/scc_to_psp.go

```diff
@@ -0,0 +1,291 @@
+func extractAllowedCapabilities(scc *securityapi.SecurityContextConstraints) []v1.Capability {
+	return convertCapabilities(scc.AllowedCapabilities)
+}
+
+// SCC.Volumes are the same as PSP.Volumes.
+func extractVolumes(scc *securityapi.SecurityContextConstraints) []extensions.FSType {
+	volumes := make([]extensions.FSType, 0, len(scc.Volumes))
+	for _, volume := range scc.Volumes {
+		fsType := extensions.FSType(string(volume))
+		volumes = append(volumes, fsType)
+	}
+	return volumes
+}
+
+// SCC.AllowHostNetwork is the same as PSP.HostNetwork.
+func extractHostNetwork(scc *securityapi.SecurityContextConstraints) bool {
+	return scc.AllowHostNetwork
+}
+
+// SCC.AllowHostPID is the same as PSP.HostPID.
+func extractHostPID(scc *securityapi.SecurityContextConstraints) bool {
+	return scc.AllowHostPID
+}
+
+// SCC.AllowHostIPC is the same as PSP.HostIPC.
+func extractHostIPC(scc *securityapi.SecurityContextConstraints) bool {
+	return scc.AllowHostIPC
+}
+
+// SCC.SELinuxContext is the same as PSP.SELinux.
+func extractSELinux(scc *securityapi.SecurityContextConstraints) extensions.SELinuxStrategyOptions {
+	var strategy extensions.SELinuxStrategy
+	var selOpts *v1.SELinuxOptions
+
+	switch scc.SELinuxContext.Type {
+	case securityapi.SELinuxStrategyRunAsAny:
+		strategy = extensions.SELinuxStrategyRunAsAny
+	case securityapi.SELinuxStrategyMustRunAs:
+		strategy = extensions.SELinuxStrategyMustRunAs
+	default:
+		panic(fmt.Sprintf("SCC %q has unknown SCC.SELinuxContext.Type: %q", scc.Name, scc.SELinuxContext.Type))
+	}
+
+	sccSelOpts := scc.SELinuxContext.SELinuxOptions
+	if sccSelOpts != nil {
+		selOpts = &v1.SELinuxOptions{
+			User:  sccSelOpts.User,
+			Role:  sccSelOpts.Role,
+			Type:  sccSelOpts.Type,
+			Level: sccSelOpts.Level,
+		}
+	}
+
+	opts := extensions.SELinuxStrategyOptions{
+		Rule:           strategy,
+		SELinuxOptions: selOpts,
+	}
+
+	return opts
+}
+
+// SCC.RunAsUser is similar to PSP.RunAsUser, with the following exceptions:
+// - it has an additional "MustRunAsRange" strategy
+// - "MustRunAs" may have a single value or a single range (while in PSP this strategy has a list of ranges only)
+// See for details: https://docs.openshift.org/3.6/architecture/additional_concepts/authorization.html#authorization-RunAsUser
+func extractRunAsUser(scc *securityapi.SecurityContextConstraints) extensions.RunAsUserStrategyOptions {
+	var strategy extensions.RunAsUserStrategy
+	var ranges []extensions.IDRange
+
+	switch scc.RunAsUser.Type {
+	case securityapi.RunAsUserStrategyRunAsAny:
+		strategy = extensions.RunAsUserStrategyRunAsAny
+	case securityapi.RunAsUserStrategyMustRunAs, securityapi.RunAsUserStrategyMustRunAsRange:
+		strategy = extensions.RunAsUserStrategyMustRunAs
+	case securityapi.RunAsUserStrategyMustRunAsNonRoot:
+		strategy = extensions.RunAsUserStrategyMustRunAsNonRoot
+	default:
+		panic(fmt.Sprintf("SCC %q has unknown SCC.RunAsUser.Type: %q", scc.Name, scc.RunAsUser.Type))
```

Comment:
Is panic() the best way to handle this ?
I would think it be best to emit errors and aggregate them so you can get all errors about all SCCs that are wonky at once instead of having to "somehow" fix them to be able to proceed ?

Reply:
Sounds like you were reviewing the old sources because I've implemented this today.